src/shared/dissect-image.c

```diff
@@ -841,74 +846,91 @@ static int dissect_image(
 
         uint64_t diskseq = m->loop ? m->loop->diskseq : 0;
 
-        if (verity && iovec_is_set(&verity->root_hash)) {
-                sd_id128_t fsuuid, vuuid;
+        /* If input is a single already open verity device and the filesystem type is pinned to a single
+         * fstype, and it's a single-image filesystem, then skip all the expensive probing and just shortcut
+         * to mount the dm device directly */
+        if ((flags & DISSECT_IMAGE_NO_BLKID)) {
+                root_fstype = image_policy_get_exhaustively(policy, PARTITION_ROOT);
+                if (root_fstype < 0)
+                        return root_fstype;
+
+                root_fstype &= _PARTITION_POLICY_FSTYPE_SET_MASK;
+                if (!(verity && verity->data_path &&
+                                (flags & DISSECT_IMAGE_NO_PARTITION_TABLE) &&
+                                (root_fstype & _PARTITION_POLICY_FSTYPE_SET_MASK) &&
+                                popcount((unsigned long long)root_fstype) == 1  && /* It's a power of 2 -> only a single bit is set -> single fstype pinning */
+                                path_startswith(devname, "/dev/mapper/")))
+                        return -ENOPKG;
```

Comment:
hmm, i'd prefer if we'd just call partition_policy_flags_to_string() here, and then rely on it's return value (as suggested further up) to determine if more than one flag is set. (instead of using popcount)